### ISSUE TITLE
[Bug]: Capacitor on Android detected as web platform after redirecting with `allowNavigation`

### ISSUE THREAD

**Issue by Dylancyclone** [ionic-team/capacitor#7454]
### Capacitor Version

💊   Capacitor Doctor  💊 

Latest Dependencies:

  @capacitor/cli: 6.0.0
  @capacitor/core: 6.0.0
  @capacitor/android: 6.0.0
  @capacitor/ios: 6.0.0

Installed Dependencies:

  @capacitor/cli: 6.0.0
  @capacitor/core: 6.0.0
  @capacitor/ios: 6.0.0
  @capacitor/android: 6.0.0

[success] iOS looking great! 👌
[success] Android looking great! 👌

### Other API Details

```Shell
`npm --version`
> 10.7.0

`node --version`
> v20.11.0
```

### Platforms Affected

- [ ] iOS
- [X] Android
- [ ] Web

### Current Behavior

After redirecting to an external URL that is allowed in the `server.allowNavigation` config setting, an android app reports being on a web platform instead of android.
This can be seen by calling `Capacitor.getPlatform()` (which returns `android` before redirecting and `web` after redirecting) or by trying to use a native API, which will either fallback to the web version of the API, or fail to run at all

All this works on iOS, where after redirecting, the app can still use native iOS APIs

### Expected Behavior

After redirecting, the android app should still have access to all the native Android APIs (and report itself as being an Android platform)

### Project Reproduction

https://github.com/Dylancyclone/capacitor-android-redirect/tree/main

### Additional Information

Reproduction details are in the repo, but the basic reproduction steps are as follows:

1. Serve a valid capacitor app on your computer (i.e. with `npx serve .`)
2. Add your computer's IP to the `server.allowNavigation` capacitor config setting
3. Load the application on your (real or simulated) device
4. Notice that `Capacitor.getPlatform()` returns `android` on android and `ios` on iOS
5. Redirect to the hosted server (any method works, can be `window.location.assign()`, `window.location.href = ""`, `window.location.replace()`, or redirecting through a form
6. Notice that `Capacitor.getPlatform()` now returns `web` on android but still `ios` on iOS

Possibly related to https://github.com/ionic-team/capacitor/issues/5455

Here are videos showing the issue. Even though these were recorded on a simulator, the results are identical to a physical device

https://github.com/ionic-team/capacitor/assets/19830705/076fc893-1732-4bab-9aea-b7a90b4ec142

https://github.com/ionic-team/capacitor/assets/19830705/4522dcef-bbdb-42e4-8085-7009f55176bf

Our use case is to have one app that connects to different backends on different subdomains, very similar to Slack. So far everything is working except the android app loses access to its native APIs

**Comment by Dylancyclone** [comment]
So it looks like there are two parts to this issue: redirecting to a local address, and capacitor not injecting the JS bridge into external pages. Again, both of these are only relevant to Android, the iOS side works perfectly

For the first issue, if we try to navigate to a computer on the local network (http, through ip address or `[computername].local`) we would get the issue described above where the platform would be seen as `web` and the plugins would use the web version or fail entirely. If we instead redirect to a remotely hosted server (https), we get a slightly different issue.

If we add our remote server to the `server.allowNavigation` config and redirect to it using the steps above, the platform is reported as `android` correctly, but all plugins are completely broken:

![image](https://github.com/ionic-team/capacitor/assets/19830705/d2a0d8cb-68e3-4a4a-b75a-9ac12dc65c8a)

Investigating this issue, we found the root cause to be the JavaScript injection in [Bridge.java](https://github.com/ionic-team/capacitor/blob/main/android/capacitor/src/main/java/com/getcapacitor/Bridge.java). We notice that if `WebViewFeature.DOCUMENT_START_SCRIPT`is supported, the javascript is only injected into the base URL (if it is NOT supported (i.e by commenting the if guard out), the injector always works). If the `addDocumentStartJavaScript()` function were changed to inject on all `allowedOriginRules` (the base URL and everything in the `server.allowNavigation` setting) instead of just the base URL, all plugins work again.

Here's the diff:

https://github.com/ionic-team/capacitor/blob/f8264ccae1f2fec553521bc62d767c4909ea6d79/android/capacitor/src/main/java/com/getcapacitor/Bridge.java#L254-L269

```java
        // Start the local web server
        JSInjector injector = getJSInjector();
        if (WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) {
            WebViewCompat.addDocumentStartJavaScript(webView, injector.getScriptString(), allowedOriginRules);
            injector = null;
        }
        localServer = new WebViewLocalServer(context, this, injector, authorities, html5mode);
```

I am new to capacitor and have no idea if this would have any side effects, but this seems to fix our issue completely and bring parity with the iOS version.

<hr>

So to summarize:

The original issue still exists while redirecting to a computer on a local network (or just over http, I don't have the capability to test that right now)

There also appears to be a bug when redirecting to a remote server (again, maybe just over https) caused by the native bridge not injecting capacitor's Javascript on remote pages. A fix we found was to make the change above, but I'd rather ask if it is the correct approach before creating a pull request.

**Comment by jarredhawkins** [comment]
> I have hacked together a quick "patch" that we call in the MainActivity, to avoid editing the Bridge.Java file directly (as it gets messy with version control). Here it is in case anyone wants to use it (call it after `onCreate` and before loading any other pages):

@Vadinci huge thanks for that snippet. I hacked together similar in java:

```java
private void patchJSInjection() {
        try {
            if (WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) {
                var getJsInjector = Arrays.stream(bridge.getClass().getDeclaredMethods())
                        .filter(method -> method.getName().equals("getJSInjector"))
                        .findFirst()
                        .get();

                getJsInjector.setAccessible(true);
                var injector = getJsInjector.invoke(bridge);

                var getScriptString = Arrays.stream(injector.getClass().getDeclaredMethods())
                        .filter(method -> method.getName().equals("getScriptString"))
                        .findFirst()
                        .get();
                var scriptString = (String) getScriptString.invoke(injector);

                var allowedOrigins = Arrays.stream(bridge.getConfig().getAllowNavigation())
                        .filter(str -> str.contains("yourdomain.com") || str.contains("otherdomain.com"))
                        .filter(str -> str.contains("https://"))
                        // WebViewCompat likes things formatted particularly, trim trailing /*
                        .map(str -> str.replaceAll("/\\*$", ""))
                        .collect(Collectors.toSet());

                Logger.info("patchJSInjection", "Injecting custom rules " + allowedOrigins);
                WebViewCompat.addDocumentStartJavaScript(bridge.getWebView(), scriptString, allowedOrigins);
            }
        } catch (Exception e) {
            Logger.error( e.getMessage(), e);
        }
}
```

**Comment by Cow258** [comment]
To get around this issue, I just patch the `Bridge.java` after built

```js
const xfs = require('fs/promises')

const { glob } = require('glob')

module.exports = async () => {

  const originPatten = `\
        // Start the local web server
        JSInjector injector = getJSInjector();
        if (WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) {
            String allowedOrigin = appUrl;
            Uri appUri = Uri.parse(appUrl);
            if (appUri.getPath() != null) {
                if (appUri.getPath().equals("/")) {
                    allowedOrigin = appUrl.substring(0, appUrl.length() - 1);
                } else {
                    allowedOrigin = appUri.toString().replace(appUri.getPath(), "");
                }
            }
            WebViewCompat.addDocumentStartJavaScript(webView, injector.getScriptString(), Collections.singleton(allowedOrigin));
            injector = null;
        }
        localServer = new WebViewLocalServer(context, this, injector, authorities, html5mode);`.split('\n')
  const replacePatten = `\
        // Patched local web server
        JSInjector injector = getJSInjector();
        if (WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) {
            WebViewCompat.addDocumentStartJavaScript(webView, injector.getScriptString(), allowedOriginRules);
            injector = null;
        }
        localServer = new WebViewLocalServer(context, this, injector, authorities, html5mode);`.split('\n')

  let modified = false
  for (const path of await glob('node_modules/.pnpm/**/Bridge.java')) {
    const source = await xfs.readFile(path, 'utf8')
    const lines = source.split('\n')

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]

      if (line.includes('// Start the local web server')) {
        if (lines.slice(i, i + originPatten.length).join('\n') !== originPatten.join('\n')) {
          console.log('Skipped: Pattern not matched')
          continue
        }

        console.log('Modified: Pattern matched')
        lines.splice(i, originPatten.length, ...replacePatten)
        modified = true
        break
      }
    }

    if (modified) {
      await xfs.writeFile(path, lines.join('\n'), 'utf8')
      break
    }
  }
}

```

**Comment by rfe-css** [comment]
In our case the issue was, that the script did not get injected on allowNavigation pages because it is not provided to the addDocumentStartJavaScript function:

![image](https://github.com/user-attachments/assets/c181f1f9-d048-4599-b06d-67f4720055f6)

i also recommend using https://www.npmjs.com/package/patch-package for all these patches until it is being resolved.

**Comment by martin-braun** [comment]
@rfe-css Thank god. We use an initial local index.html followed by a navigation to "allowedNavigation" URLs, yet it wasn't working. The old service loader fix from #5278 isn't load either, because it is not really used, anymore.

This here is poorly implemented and not properly tested. How did v6 can be considered stable when not even PWAs are being tested?

**Comment by schniper** [comment]
> We are experiencing this issue as well, and I cannot thank you enough for having found this workaround 🙇 I can confirm that changing the `addDocumentStartJavaScript` call to just use `allowedOriginRules` fixes the issues for https requests on local networks. Using the fallback `jsInjector` does not however.
> 
> I have hacked together a quick "patch" that we call in the MainActivity, to avoid editing the Bridge.Java file directly (as it gets messy with version control). Here it is in case anyone wants to use it (call it after `onCreate` and before loading any other pages):
> 
> ```kotlin
> class MainActivity : BridgeActivity() {
>     override fun onCreate(savedInstanceState: Bundle?) {
>         super.onCreate(savedInstanceState)
>         /**
>          * ...
>          */
>         patchJSInjection();
>     }
> 
>     /**
>      * ...
>      */
> 
>     private fun patchJSInjection() {
>         try {
>             if (WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) {
>                 val getJsInjector = bridge::class.java.declaredMethods.single { it.name == "getJSInjector" };
>                 getJsInjector.isAccessible = true;
>                 val injector = getJsInjector.invoke(bridge);
> 
>                 val getScriptString = injector::class.java.declaredMethods.single { it.name === "getScriptString" };
>                 val scriptString = getScriptString.invoke(injector) as String;
> 
>                 val allowedOrigins: MutableSet<String> = mutableSetOf();
>                 // Add origins that the Capacitor JS Bridge should be injected into
>                 allowedOrigins.add("https://www.foo.bar");
> 
>                 WebViewCompat.addDocumentStartJavaScript(bridge.webView, scriptString, allowedOrigins)
>             }
>         }catch (e: Exception) {
>             Log.e("Error", e.message ?: "");
>         }
>     }
> }
> ```

Thank you so much for this. I was thinking it had something to do with the JS code.
One mention: minSdkVersion will need to get bumped to 24, from the default 22. Just for the Java syntax sugar.

**Comment by schniper** [comment]
> > I have hacked together a quick "patch" that we call in the MainActivity, to avoid editing the Bridge.Java file directly (as it gets messy with version control). Here it is in case anyone wants to use it (call it after `onCreate` and before loading any other pages):
> 
> @Vadinci huge thanks for that snippet. I hacked together similar in java:
> 
> ```java
> private void patchJSInjection() {
>         try {
>             if (WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) {
>                 var getJsInjector = Arrays.stream(bridge.getClass().getDeclaredMethods())
>                         .filter(method -> method.getName().equals("getJSInjector"))
>                         .findFirst()
>                         .get();
> 
>                 getJsInjector.setAccessible(true);
>                 var injector = getJsInjector.invoke(bridge);
> 
>                 var getScriptString = Arrays.stream(injector.getClass().getDeclaredMethods())
>                         .filter(method -> method.getName().equals("getScriptString"))
>                         .findFirst()
>                         .get();
>                 var scriptString = (String) getScriptString.invoke(injector);
> 
>                 var allowedOrigins = Arrays.stream(bridge.getConfig().getAllowNavigation())
>                         .filter(str -> str.contains("yourdomain.com") || str.contains("otherdomain.com"))
>                         .filter(str -> str.contains("https://"))
>                         // WebViewCompat likes things formatted particularly, trim trailing /*
>                         .map(str -> str.replaceAll("/\\*$", ""))
>                         .collect(Collectors.toSet());
> 
>                 Logger.info("patchJSInjection", "Injecting custom rules " + allowedOrigins);
>                 WebViewCompat.addDocumentStartJavaScript(bridge.getWebView(), scriptString, allowedOrigins);
>             }
>         } catch (Exception e) {
>             Logger.error( e.getMessage(), e);
>         }
> }
> ```

Thank you so much for this. I was thinking it had something to do with the JS code.
One mention: minSdkVersion will need to get bumped to 24, from the default 22. Just for the Java syntax sugar.

**Comment by arieftb** [comment]
> I face this issue too when navigate to different sub domain.
> 
> Hope this issue can be fixed.
> 
> Is patchJSInjection solution still work ?

OK the patchJSInjection solution still work. Thanks

**Comment by rally25rs** [comment]
The write-up in this issue seems to imply that this issue only happens with `allowNavigation` and changing locations, but I see this behavior without `allowNavigation`, using the basic config:

```
const config: CapacitorConfig = {
  appId: "...",
  appName: "...",
  webDir: "dist",
  loggingBehavior: "production",
  server: {
    cleartext: true,
  },
  plugins: {
    CapacitorHttp: {
      enabled: true,
    },
  },
};
```

Using `@capacitor/core@7.1.0` and the Ionic VSCode plugin: Build / Sync / Run Android.
When the app opens in the android emulator, `console.log(Capacitor.getPlatform());` prints `"web"`.

**Comment by adrianomarquess** [comment]
I had the same problem in a production app, this solution worked for me 🙌🏾

```java
import android.os.Bundle;
import android.os.PersistableBundle;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.webkit.WebViewCompat;
import androidx.webkit.WebViewFeature;
import com.getcapacitor.BridgeActivity;
import java.lang.reflect.Method;
import java.util.HashSet;
import java.util.Set;

public class MainActivity extends BridgeActivity {
    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        patchJSInjection();
    }

    private void patchJSInjection() {
        try {
            if (WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) {
                Method getJsInjector = bridge.getClass().getDeclaredMethod("getJSInjector");
                getJsInjector.setAccessible(true);
                Object injector = getJsInjector.invoke(bridge);

                assert injector != null;
                Method getScriptString = injector.getClass().getDeclaredMethod("getScriptString");
                String scriptString = (String) getScriptString.invoke(injector);

                Set<String> allowedOrigins = new HashSet<>();
                // Add origins that the Capacitor JS Bridge should be injected into
                allowedOrigins.add("https://foo.example.com");
                allowedOrigins.add("https://bar.example.com");

                assert scriptString != null;
                WebViewCompat.addDocumentStartJavaScript(bridge.getWebView(), scriptString, allowedOrigins);
            }
        } catch (Exception e) {
            Log.e("Error", e.getMessage() != null ? e.getMessage() : "");
        }
    }
}
